src/java_bytecode/java_bytecode_language.cpp

```diff
@@ -478,53 +478,53 @@ bool java_bytecode_languaget::typecheck(
   // Must load java.lang.Object first to avoid stubbing
   // This ordering could alternatively be enforced by
   // moving the code below to the class loader.
-  java_class_loadert::class_mapt::const_iterator it=
-    java_class_loader.class_map.find("java.lang.Object");
-  if(it!=java_class_loader.class_map.end())
+  java_class_loadert::parse_tree_with_overridest_mapt::const_iterator it =
+    java_class_loader.get_class_with_overlays_map().find("java.lang.Object");
+  if(it != java_class_loader.get_class_with_overlays_map().end())
   {
     if(java_bytecode_convert_class(
-     it->second,
-     symbol_table,
-     get_message_handler(),
-     max_user_array_length,
-     method_bytecode,
-     lazy_methods_mode,
-     string_preprocess))
-       return true;
+         it->second,
+         symbol_table,
+         get_message_handler(),
+         max_user_array_length,
+         method_bytecode,
+         lazy_methods_mode,
+         string_preprocess))
+    {
+      return true;
+    }
   }
 
   // first generate a new struct symbol for each class and a new function symbol
   // for every method
-  for(java_class_loadert::class_mapt::const_iterator
-      c_it=java_class_loader.class_map.begin();
-      c_it!=java_class_loader.class_map.end();
-      c_it++)
+  for(const auto &class_trees : java_class_loader.get_class_with_overlays_map())
   {
-    if(c_it->second.parsed_class.name.empty())
+    if(class_trees.second.front().parsed_class.name.empty())
       continue;
 
-    if(
-      java_bytecode_convert_class(
-        c_it->second,
-        symbol_table,
-        get_message_handler(),
-        max_user_array_length,
-        method_bytecode,
-        lazy_methods_mode,
-        string_preprocess))
+    if(java_bytecode_convert_class(
+         class_trees.second,
+         symbol_table,
+         get_message_handler(),
+         max_user_array_length,
+         method_bytecode,
+         lazy_methods_mode,
+         string_preprocess))
+    {
       return true;
+    }
   }
 
   // find and mark all implicitly generic class types
   // this can only be done once all the class symbols have been created
-  for(const auto &c : java_class_loader.class_map)
+  for(const auto &c : java_class_loader.get_class_with_overlays_map())
   {
-    if(c.second.parsed_class.name.empty())
+    if(c.second.front().parsed_class.name.empty())
       continue;
     try
     {
       mark_java_implicitly_generic_class_type(
-        c.second.parsed_class.name, symbol_table);
+        c.second.front().parsed_class.name, symbol_table);
```

Comment:
Is there a reason we only care about the first parse tree in this case? Note an implicitly generic class means something like:

```java
class Outer<T> {
 class Inner { 
    T field;
  }
}
```

Inner isn't explicitly generic (has no generic parameters), but clearly is in some sense a generic type as one of its fields is of type `T`. I suspect you are better placed to know if it is important for any overlayed class parse trees to be correctly, but my feeling is it probably should be (the parse essentially means that the type of `field` is correctly updated to be `Outer::T`)

Reply:
We care about them all, but they all have the same name. So `c.second.front().parsed_class.name` should be the same as `c.second[x].parsed_class.name`. At least that was my reasoning. I may be wrong.

Reply:
Ah sorry obviously didn't read it carefully enough - you are of course correct here.